src/error.rs

```diff
@@ -5,78 +5,119 @@
 // <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
-use core::convert::From;
 use core::fmt;
 use core::num::NonZeroU32;
 
-// A randomly-chosen 24-bit prefix for our codes
-pub(crate) const CODE_PREFIX: u32 = 0x57f4c500;
-const CODE_UNKNOWN: u32 = CODE_PREFIX | 0x00;
-const CODE_UNAVAILABLE: u32 = CODE_PREFIX | 0x01;
-
-/// The error type.
+/// A small and `no_std` compatible error type. It can indicate failure from
+/// either the underlying OS or a custom error reason.
 ///
-/// This type is small and no-std compatible.
+/// The [`Error::raw_os_error()`] will indicate if the error is from the OS, and
+/// if so, which error code the OS gave the application. If such an error is
+/// encountered, please consult with your system documentation.
 #[derive(Copy, Clone, Eq, PartialEq)]
-pub struct Error(pub(crate) NonZeroU32);
+pub struct Error(NonZeroU32);
 
+// This NonZeroU32 has enough room to store 3 types of values:
+//   - OS Errors:         in range [1, 1 << 31) (i.e. positive i32 values)
+//   - Custom u16 Errors: in range [1 << 31, 1 << 31 + 1 << 16)
+//   - Unknown Errors:    currently just (1 << 32) - 1
+// TODO simplify impls with try_from when version >= 1.34
 impl Error {
     /// An unknown error.
-    pub const UNKNOWN: Error = Error(unsafe { NonZeroU32::new_unchecked(CODE_UNKNOWN) });
+    pub(crate) const UNKNOWN: Error = Self(unsafe { NonZeroU32::new_unchecked(u32::max_value()) });
+    const CUSTOM_START: u32 = 1 << 31;
 
-    /// No generator is available.
-    pub const UNAVAILABLE: Error = Error(unsafe { NonZeroU32::new_unchecked(CODE_UNAVAILABLE) });
+    pub(crate) fn from_os_error(errno: i32) -> Self {
+        if errno > 0 {
+            Self(NonZeroU32::new(errno as u32).unwrap())
+        } else {
+            Self::UNKNOWN
+        }
+    }
 
-    /// Extract the error code.
-    ///
-    /// This may equal one of the codes defined in this library or may be a
-    /// system error code.
-    ///
-    /// One may attempt to format this error via the `Display` implementation.
-    pub fn code(&self) -> NonZeroU32 {
-        self.0
+    pub(crate) fn from_custom_error(custom: u16) -> Self {
+        Self(NonZeroU32::new(custom as u32 + Self::CUSTOM_START).unwrap())
```

Comment:
If we use a thin wrapper around this for `rand_core::Error` then we need some public way of constructing errors with custom codes. Maybe we should even use a pair: `block: u16, code: u16`?

Reply:
Fixed though use of the `from_kind()` method to mirror the `kind()` method. Now users can create their own errors like:
```rust
const MY_BLOCK: i16 = 2;
const MY_ERROR: u16 = 200;
Error::from_kind(ErrorKind::Custom(MY_BLOCK, MY_ERROR)
```